### ISSUE TITLE
[RFC] CPUID Enumeration

### ISSUE THREAD

**Issue by rianquinn** [Bareflank/hypervisor#849]
TL;DR

Bareflank should only ack using CPUID 0x4BF00000. Downstream projects will use CPUID 0x40000000-0x40010000.

Description:

We need the ability for Linux to detect the presence of Bareflank without colliding with downstream projects (for example Boxy). If Bareflank is used without mods, Linux still needs some custom Bareflank specific mods to deal with INIT/SIPI issues properly. We also have to use CPUID as it is the only instruction that can be executed in userspace and the kernel without generating invalid instruction exceptions (which means that MSRs, VMcalls, and Port IO are all off limits). 

The Linux kernel uses CPUID 0x40000000 to detect the presence of a hypervisor (although I don't think VMWare uses this method). How this works is in your detect() function, you return the CPUID leaf that was detected. 
https://elixir.bootlin.com/linux/v4.19.34/source/arch/x86/kernel/cpu/hypervisor.c#L63

Some hypervisors like HyperV have this hardcoded, but Xen and KVM use this function:
https://elixir.bootlin.com/linux/v4.19.34/source/arch/x86/include/asm/processor.h#L959

Right now, Bareflank uses 0x4BF00000 to detect Bareflank which is perfect. The reason this works so well is the above function will return false as it is not within the range of detection and 0x40000000 is still open for downstream projects to implement, including using the above function for detection. 

No mods are needed to Bareflank to support this. Linux, however, will need a Bareflank specific detection mechanism that is implemented like this:
https://elixir.bootlin.com/linux/v4.19.34/source/arch/x86/kernel/kvm.c#L670

But in reverse. First, Bareflank will ack itself using 0x4BF00000. If Linux doesn't see a proper ack, it will continue on without doing anything else. If a proper ack is seen, it will first loop through these leaves and if all 0s are returned, it will assume that no downstream projects are being used, in which case Bareflank itself is detected. If however a downstream project is detected, it will return not detected so that the downstream project takes priority. 

Using this method, we should be able to:
- Handle INIT/SIPI issues with versions of Linux that have the delay turned off like Arch Linux when only Bareflank is being used. 
- Provide a means for downstream projects to implement their own enumeration techniques using the already provided mechanisms that Linux supports
- Ensure the design has a minimal impact on Linux so that projects like Ubuntu can leave Bareflank support enabled by default.

**Comment by connojd** [comment]
:+1:
